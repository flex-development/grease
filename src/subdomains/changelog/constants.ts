/**
 * @file Constants
 * @module grease/changelog/constants
 */

import type { ICommitType } from '#src/changelog/interfaces'
import { Type } from '@flex-development/commitlint-config'

/**
 * An array of objects representing explicitly supported commit message types,
 * and whether they should show up in the changelog.
 *
 * @see {@linkcode ICommitType}
 *
 * @const {ICommitType[]} TYPES
 */
export const TYPES: ICommitType[] = [
  { section: ':sparkles: Features', type: Type.FEAT },
  { section: ':bug: Fixes', type: Type.FIX },
  { section: ':mechanical_arm: Refactors', type: Type.REFACTOR },
  { section: ':fire: Performance Improvements', type: Type.PERF },
  { section: ':wastebasket: Reverts', type: Type.REVERT },
  { section: ':pencil: Documentation', type: Type.DOCS },
  { section: ':white_check_mark: Testing', type: Type.TEST },
  { section: ':package: Build', type: Type.BUILD },
  { section: ':robot: Continuous Integration', type: Type.CI },
  { section: ':house_with_garden: Housekeeping', type: Type.CHORE },
  { hidden: true as const, type: Type.RELEASE },
  { hidden: true as const, type: Type.STYLE },
  { hidden: true as const, type: Type.WIP }
]
