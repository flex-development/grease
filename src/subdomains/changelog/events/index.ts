/**
 * @file Entry Point - Events
 * @module grease/changelog/events
 */

export { default as ChangelogEvent } from './changelog.event'
export { default as ChangelogEventListener } from './changelog.listener'
