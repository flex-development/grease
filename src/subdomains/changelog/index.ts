/**
 * @file Entry Point - changelog
 * @module grease/changelog
 */

export { default as ChangelogModule } from './changelog.module'
export * from './constants'
export * from './events'
export type * from './interfaces'
export * from './models'
export * from './operations'
export * from './queries'
export type * from './types'
