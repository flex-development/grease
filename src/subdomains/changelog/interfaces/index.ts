/**
 * @file Entry Point - Interfaces
 * @module grease/changelog/interfaces
 */

export type { default as IChangelogEntry } from './changelog-entry.interface'
export type { default as ICommitGroup } from './commit-group.interface'
export type { default as ICommitType } from './commit-type.interface'
