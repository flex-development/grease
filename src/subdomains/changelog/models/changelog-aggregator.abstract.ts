/**
 * @file Models - AbstractChangelogAggregator
 * @module grease/changelog/models/AbstractChangelogAggregator
 */

import type { BreakingChange, Commit, Mention, Reference } from '#src/git'
import {
  hasOwn,
  select,
  type CompareResult,
  type PropertyKey
} from '@flex-development/tutils'
import type CommitGroup from './commit-group.model'
import type CommitType from './commit-type.model'

/**
 * Abstract changelog aggregator.
 *
 * @template T - Parsed commit type
 *
 * @abstract
 * @class
 */
abstract class AbstractChangelogAggregator<T extends Commit = Commit> {
  /**
   * Parsed commits.
   *
   * @public
   * @readonly
   * @instance
   * @member {ReadonlyArray<T>} commits
   */
  public readonly commits: readonly T[]

  /**
   * An array of objects representing explicitly supported commit message types,
   * and whether they should show up in the changelog.
   *
   * @see {@linkcode CommitType}
   *
   * @protected
   * @readonly
   * @instance
   * @member {ReadonlyArray<CommitType>} types
   */
  protected readonly types: readonly CommitType[]

  /**
   * Create a new changelog aggregator.
   *
   * @param {ReadonlyArray<CommitType>} types - Commit message types
   * @param {ReadonlyArray<T>?} [commits=[]] - Parsed commits
   */
  constructor(types: readonly CommitType[], commits: readonly T[] = []) {
    this.commits = Object.freeze(commits)
    this.types = Object.freeze(select(types, t => hasOwn(this.map, t.type)))
  }

  /**
   * Get a breaking changes array.
   *
   * @see {@linkcode BreakingChange}
   *
   * @public
   * @abstract
   *
   * @return {BreakingChange[]} Breaking changes array
   */
  public abstract get breaks(): BreakingChange[]

  /**
   * Get a sorted commit groups array.
   *
   * @see {@linkcode CommitGroup}
   *
   * @public
   * @abstract
   *
   * @return {CommitGroup<T>[]} Sorted commit groups array
   */
  public abstract get groups(): CommitGroup<T>[]

  /**
   * Get a parsed commit map.
   *
   * @protected
   * @abstract
   *
   * @return {Record<string, T[]>} Parsed commit map
   */
  protected abstract get map(): Record<string, T[]>

  /**
   * Get a mentions array.
   *
   * @see {@linkcode Mention}
   *
   * @public
   * @abstract
   *
   * @return {Mention[]} Mentions array
   */
  public abstract get mentions(): Mention[]

  /**
   * Get an array containing issue and pull request references.
   *
   * @see {@linkcode Reference}
   *
   * @public
   * @abstract
   *
   * @return {Reference[]} References array
   */
  public abstract get references(): Reference[]

  /**
   * Get a commit group key.
   *
   * @protected
   * @abstract
   *
   * @param {T} commit - Parsed commit
   * @return {PropertyKey} Commit group key
   */
  protected abstract groupby(commit: T): PropertyKey

  /**
   * Breaking change sorter.
   *
   * @see {@linkcode BreakingChange}
   *
   * @protected
   * @abstract
   *
   * @param {BreakingChange} brk1 - First breaking change to compare
   * @param {BreakingChange} brk2 - Second breaking change to compare
   * @return {CompareResult} Comparison result
   */
  protected abstract sortbrk(
    brk1: BreakingChange,
    brk2: BreakingChange
  ): CompareResult

  /**
   * Commit sorter.
   *
   * @protected
   * @abstract
   *
   * @param {T} cmt1 - First commit to compare
   * @param {T} cmt2 - Second commit to compare
   * @return {CompareResult} Comparison result
   */
  protected abstract sortcmt(cmt1: T, cmt2: T): CompareResult

  /**
   * Commit group sorter.
   *
   * @see {@linkcode CommitGroup}
   *
   * @protected
   * @abstract
   *
   * @param {CommitGroup<T>} g1 - First commit group to compare
   * @param {CommitGroup<T>} g2 - Second commit group to compare
   * @return {CompareResult} Comparison result
   */
  protected abstract sortgrp(
    g1: CommitGroup<T>,
    g2: CommitGroup<T>
  ): CompareResult
}

export default AbstractChangelogAggregator
