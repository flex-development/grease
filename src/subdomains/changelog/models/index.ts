/**
 * @file Entry Point - Models
 * @module grease/changelog/models
 */

export {
  default as AbstractChangelogAggregator
} from './changelog-aggregator.abstract'
export { default as ChangelogAggregator } from './changelog-aggregator.model'
export {
  default as ChangelogEntry,
  type ChangelogEntryDTO
} from './changelog-entry.model'
export {
  default as AbstractChangelogFormatter
} from './changelog-formatter.abstract'
export { default as ChangelogFormatter } from './changelog-formatter.model'
export { default as ChangelogInfile } from './changelog-infile.model'
export {
  default as ChangelogStream,
  type ChangelogStreamDTO
} from './changelog-stream.model'
export { default as CommitGroup } from './commit-group.model'
export { default as CommitType } from './commit-type.model'
