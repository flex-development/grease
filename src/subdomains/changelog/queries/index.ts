/**
 * @file Entry Point - Queries
 * @module grease/changelog/queries
 */

export { default as ChangelogQueryHandler } from './changelog.handler'
export {
  default as ChangelogQuery,
  type ChangelogQueryParams
} from './changelog.query'
