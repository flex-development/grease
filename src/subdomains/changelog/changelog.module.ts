/**
 * @file ChangelogModule
 * @module grease/changelog/ChangelogModule
 */

import { LoggerService } from '#src/providers'
import { Module } from '@nestjs/common'
import { ChangelogEventListener } from './events'
import { ChangelogOperationHandler } from './operations'
import { ChangelogQueryHandler } from './queries'

/**
 * Changelog module.
 *
 * @class
 */
@Module({
  providers: [
    ChangelogEventListener,
    ChangelogOperationHandler,
    ChangelogQueryHandler,
    LoggerService
  ]
})
class ChangelogModule {}

export default ChangelogModule
