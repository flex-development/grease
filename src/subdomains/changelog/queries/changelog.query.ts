/**
 * @file Queries - ChangelogQuery
 * @module grease/changelog/ChangelogQuery
 */

import * as constants from '#src/changelog/constants'
import type { ICommitType } from '#src/changelog/interfaces'
import { ChangelogAggregator, CommitType } from '#src/changelog/models'
import { IsConstructor } from '#src/decorators'
import { GitCommitOptions, type Commit } from '#src/git'
import {
  defaults,
  define,
  fallback,
  select,
  unique,
  type Assign,
  type Partial
} from '@flex-development/tutils'
import {
  ArrayNotEmpty,
  IsArray,
  IsBoolean,
  IsInstance,
  IsNumber,
  Min,
  ValidateNested
} from 'class-validator'

/**
 * Changelog entry query parameters.
 *
 * @template T - Parsed commit type
 */
type ChangelogQueryParams<T extends Commit = Commit> = Assign<
  Partial<ChangelogQuery<T>>,
  {
    /**
     * Changelog aggregator.
     *
     * @see {@linkcode ChangelogAggregator}
     */
    Aggregator?: typeof ChangelogAggregator<T>

    /**
     * An array of commit group objects representing explicitly supported commit
     * message types, and whether they should show up in the changelog entry.
     *
     * @see {@linkcode ICommitType}
     *
     * @default constants.TYPES
     */
    types?: ICommitType[]
  }
>

/**
 * Changelog entries query.
 *
 * @see {@linkcode GitCommitOptions}
 *
 * @template T - Parsed commit type
 *
 * @class
 * @extends {GitCommitOptions<T>}
 */
class ChangelogQuery<T extends Commit = Commit> extends GitCommitOptions<T> {
  /**
   * Changelog aggregator class.
   *
   * @default ChangelogAggregator
   *
   * @public
   * @instance
   * @member {typeof ChangelogAggregator<T>} Aggregator
   */
  @IsConstructor()
  public Aggregator: typeof ChangelogAggregator<T>

  /**
   * Number of entries to query.
   *
   * If `0`, all entries are queried.
   *
   * If at least `0`, the commit revision range start ({@linkcode from}) will
   * also be overridden based on this value when the query is executed.
   *
   * @default 1
   *
   * @public
   * @instance
   * @member {number} releases
   */
  @IsNumber()
  @Min(-1)
  public releases: number

  /**
   * An array of objects representing explicitly supported commit message types
   * and their visibility in the changelog.
   *
   * @see {@linkcode CommitType}
   *
   * @default
   *  select(constants.TYPES, null, type => new CommitType(type))
   *
   * @public
   * @instance
   * @member {CommitType[]} types
   */
  @IsArray()
  @ArrayNotEmpty()
  @IsInstance(CommitType, { each: true })
  @ValidateNested()
  public types: CommitType[]

  /**
   * Include unstable tags.
   *
   * @default true
   *
   * @public
   * @instance
   * @member {boolean} unstable
   */
  @IsBoolean()
  public unstable: boolean

  /**
   * Create a new changelog entries query.
   *
   * @see {@linkcode ChangelogQueryParams}
   *
   * @param {ChangelogQueryParams<T>?} [params] - Query parameters
   */
  constructor(params?: ChangelogQueryParams<T>) {
    super(params)

    const {
      Aggregator,
      releases,
      types,
      unstable
    } = defaults(fallback(params, {}), {
      Aggregator: ChangelogAggregator,
      releases: 1,
      types: constants.TYPES,
      unstable: true
    })

    this.Aggregator = Aggregator
    this.releases = releases
    this.types = unique(select(types, null, type => new CommitType(type)))
    this.unstable = unstable

    // reset revision range if all entries are queried
    !this.releases && define(this, 'from', { value: '' })
  }
}

export { ChangelogQuery as default, type ChangelogQueryParams }
