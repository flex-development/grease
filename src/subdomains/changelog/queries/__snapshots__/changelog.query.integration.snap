// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:changelog/queries/ChangelogQuery > validation > should fail validation if schema is invalid 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isConstructor": "Aggregator must be a constructor",
    },
    "property": "Aggregator",
    "value": [Function Aggregator],
  },
  ValidationError {
    "children": [],
    "constraints": {
      "min": "releases must not be less than -1",
    },
    "property": "releases",
    "value": -2,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "arrayNotEmpty": "types should not be empty",
    },
    "property": "types",
    "value": [],
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isBoolean": "unstable must be a boolean value",
    },
    "property": "unstable",
    "value": 0,
  },
]
`;
